File: solutions/264.ugly-number-ii/ugly-number-ii.js
/**
 * @param {number} n
 * @return {number}
 */
const nthUglyNumber = n => {
  let arr = [1]

  let p1 = 0
  let p2 = 0
  let p3 = 0

  for (let i = 1; i < n; i++) {
    let next = Math.min(arr[p1] * 2, arr[p2] * 3, arr[p3] * 5)
    arr[i] = next

    if (next === arr[p1] * 2) p1++
    if (next === arr[p2] * 3) p2++
    if (next === arr[p3] * 5) p3++
  }

  return arr[arr.length - 1]
}

File: solutions/264.ugly-number-ii/ugly-number-ii.py
class Solution:
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        ugly = [1]
        i2 = 0
        i3 = 0
        i5 = 0

        while n > 1:
            c2 = ugly[i2] * 2
            c3 = ugly[i3] * 3
            c5 = ugly[i5] * 5

            min_c = min(c2, c3, c5)

            if min_c == c2:
                i2 += 1

            if min_c == c3:
                i3 += 1

            if min_c == c5:
                i5 += 1

            ugly.append(min_c)
            n -= 1

        return ugly[-1]
